Use "babel-preset-react-native"

skevy opened the pull request:
Rather than specifying Babel plugins in the `.babelrc` packaged with react-native, leverage a Babel preset to define the plugins (https://github.com/exponentjs/babel-preset-react-native).

This allows for a much better user experience for those who want (or need) to override options in their project's `.babelrc`.

Prior to this PR, if a user wanted to use a custom babel-plugin (or a custom set of babel plugins), they'd have either 1) manually override the `.babelrc` in the react-packager directory (or fork RN), or 2) specify a custom transformer to use when running the packager that loaded their own `.babelrc`. Note - the custom transformer was necessary because without it, RN's `.babelrc` options would supersede the options defined in the project's `.babelrc`...potentially causing issues with plugin ordering.

This PR makes the transformer check for the existence of a project-level `.babelrc`, and if it it's there, it _doesn't_ use the react-native `.babelrc`. This prevents any oddities with Babel plugin ordering, and leaves that up to the RN user. The RN user would then just include a `.babelrc` file in their project that looks like:

```json
{
  "presets": ["react-native"]
}
```

They can then add whatever other presets or plugins that they'd like (such as the Babel Relay Plugin or a decorators plugin, for example).

**Super important note: after the fbjs changes are merged...this is really the final piece of getting Relay running OOTB in RN**

/cc @ide @vjeux

skevy commented:
@ide @vjeux Should we move `babel-preset-react-native` under the auspices of the "facebook" or "reactjs" orgs?

skevy commented:
@martinbigio and/or @davidaurelio...vjeux suggested that one of you guys review :)

satya164 commented:
Awesome!

vjeux commented:
Can you make a folder inside of this repo like we do for react-native-cli? Would be easier to manage

skevy commented:
@vjeux that seems reasonable to me. Should it be committed through the ship it bot? Or should it live in OSS only?

satya164 commented:
Any reason for the rename of '.babelrc'?

davidaurelio commented:
Guys, let me discuss this with @martinbigio, we need to sort out one thing ;-)

ide commented:
@skevy @vjeux: About moving the preset under a folder, sounds good. I'm also happy to add whoever needs to be a collaborator to the repo & npm owners list. I think we can figure that out later.

The main thing I think is important is that the babelrc.json file depends on the preset so that RN's internal configuration stays in sync with people's custom .babelrc files that are using the preset. Also the preset should follow semver, just like any other npm package.

skevy commented:
@satya164 when @ide and I were talking, we were just thinking that since the RN babelrc is handled custom anyway, via transformer.js, that it should be named with ".json" so there's not a possibility of conflict with any other babelrc's.

ide commented on packager/transformer.js:
```diff
@@ -17,15 +17,24 @@ const json5 = require('json5');
 
-function transform(src, filename, options) {
-  options = options || {};
+const projectBabelRCPath = path.resolve(__dirname, '..', '..', '..', '.babelrc');
```
Thinking about this some more, I think the babelrc under the path specified by `--projectRoots` is what we want to use. That way if you invoke the packager and point it at another directory, we'd correctly use that directory's .babelrc.

skevy replied:
I thought about that -- but what are your thoughts on if there are multiple project roots?

davidaurelio replied:
This line only works for OS and if react native is in ``${projectDirectory}/node_modules/react-native``. I’m a little bit afraid of the apocalyptic potential of this innocent path lookup ;-)

Is anybody outside of Facebook using multiple project roots? The first project root might be a saner choice. Or we can add a command line option.

skevy replied:
Let's default to the first project root, and I'll add a command line option.

satya164 commented:
@skevy Just to note, I presume some people have a `.babelrc` like this (I do),
```json
{
  "extends": "react-native/packager/react-packager/.babelrc"
}
```

Which will break after the rename. Not a big deal, but just wanted to note.

ide commented:
@satya164 Good point. We should mention it under Breaking Changes in the release notes.

satya164 commented:
@ide Yup.

davidaurelio commented on packager/babelRegisterOnly.js:
```diff
@@ -16,7 +16,7 @@ var path = require('path');
 
 function readBabelRC() {
-  var rcpath = path.join(__dirname, 'react-packager', '.babelrc');
+  var rcpath = path.join(__dirname, 'react-packager', '.babelrc.json');
```
Let’s remove the leading dot.

alvinthen commented on packager/transformer.js:
```diff
@@ -33,28 +42,22 @@ function transform(src, filename, options) {
-  });
-
-  const result = babel.transform(src, Object.assign({}, babelRC, config));
+  return Object.assign({}, babelRC, extraConfig);
```
Shouldn't `extraConfig` be `config` instead? `config` doesn't seems to be used anywhere.

skevy commented:
@martinbigio do you have any thoughts here about how to include any of the HMR stuff in babel-preset-react-native? Maybe with an option? Or is that not going to be possible because of the second transform pass?

I ask because of `resolvePlugins` -- we don't need it anymore in transformer, and in fact, don't want it. We don't want plugins to resolve relative to react-native...we want them to resolve relative to `babel-preset-react-native`, which they'll do by default just by using the plugin.

Honestly, at the moment, nothing needs to be done...as it works fine as is. But there's a possibility that there could be different versions of `transform-es2015-block-scoping` , `transform-es2015-constants`, and `transform-es2015-modules-commonjs` being used in the HMR transform step if we don't somehow make it use the same version as the babel-preset-react-native.

skevy commented:
Hey all -

Did a few things here to get this ready for merge:

1) Brought `babel-preset-react-native` INTO the RN repo, as @vjeux suggested.
2) Moved internal transform pipeline definition into the `babel-preset-react-native` pacakge, under `babel-preset-react-native/internal`. This makes it so we don't have to worry about strange versioning issues between the external transform pipeline and the internal one, and should be able to keep everything consistent.
3) Changed the `for..of` transform to be "loose", given the discussions in #5294.
4) Passed `projectRoots` to the transformer, so that we can do as @ide and @davidaurelio suggested, and try to resolve a `.babelrc` in the first project root, rather than do that weird path crap I was doing before. :)

Let me know if you all have any other comments!

martinbigio commented:
Awesome!. Another great consequence of using presets is that we could add the `package.json` to the list of files that break the packager cache.

@davidaurelio do you want to check that nothing breaks internally with this?

bestander commented:
FYI CI tests are genuinely broken https://circleci.com/gh/facebook/react-native/1734

skevy commented:
@bestander yep, this is because babel-preset-react-native isn't pushed to NPM.

skevy commented:
@bestander turns out there is some weird issue with the transform for Jest. Will fix now and push up, so that you can see passing tests :)

skevy commented:
Rebased with master, and fixed issue that was causing failing test.

You'll see a failing test here until #5601 is merged into master and I rebase.

bestander commented:
One more rebase and it should work :)

martinbigio commented on babel-preset/configs/hot.js:
```diff
@@ -0,0 +1,28 @@
+/**
```
nit: Maybe this should be named HMR as hot loading is a usecase of hot module replacement. Don't feel strong about this so feel free to leave it as is

martinbigio commented on babel-preset/configs/internal.js:
```diff
@@ -0,0 +1,16 @@
+const resolvePlugins = require('./lib/resolvePlugins');
+
+module.exports = function(options) {
+  // For future internal pipeline usage
```
thanks for updating this!

martinbigio commented on babel-preset/package.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "homepage": "https://github.com/facebook/react-native/tree/master/babel-preset/README.md",
+  "dependencies": {
+    "babel-plugin-react-transform": "2.0.0-beta1",
```
will we need to land the pr to update babel 6.3.x first, right? I think that one hasn't been merged internally. cc @davidaurelio

martinbigio commented on package.json:
```diff
@@ -108,35 +108,12 @@
-    "babel-polyfill": "^6.0.16",
-    "babel-types": "^6.1.2",
-    "babylon": "^6.1.2",
+    "babel-core": "~6.4.0",
```
Oh, you changed it here, nvm. Thanks! :)

martinbigio commented on packager/react-packager/rn-babelrc.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "presets": [ "react-native" ],
+  "plugins": []
```
+1

martinbigio commented on packager/transformer.js:
```diff
@@ -11,47 +11,90 @@
+    if (!projectBabelRCPath || !fs.existsSync(projectBabelRCPath)) {
+      babelRC = json5.parse(
+        fs.readFileSync(
+          path.resolve(__dirname, 'react-packager', 'rn-babelrc.json')));
```
nit: last paren should be on newline

skevy commented:
@martinbigio thanks for the review. Fixing the nits now.

skevy commented:
@bestander rebased.

martinbigio commented:
@facebook-github-bot import

skevy commented:
:rocket: :rocket: :rocket:

martinbigio commented:
@facebook-github-bot import

martinbigio commented:
@facebook-github-bot  import

martinbigio commented:
@mkonicek facebook-github-bot refuses to import a pr which was already imported. Any clue about how can I get around this?

bestander commented:
@martinbigio @mkonicek is away for a week...

chirag04 commented:
:+1:

vjeux commented:
This officially broke react-native for node < 4.0. We should either revert the arrow functions or crash earlier if not using node 4

<img width="602" alt="screen shot 2016-02-04 at 1 49 27 pm" src="https://cloud.githubusercontent.com/assets/197597/12830832/21c69552-cb46-11e5-8576-fd67010347b3.png">

skevy commented:
@vjeux I definitely think we should crash earlier if not using Node 4. No one should be using less than Node 4, and it already says at the top of your screenshot that React Native runs on Node 4.0 or newer.

skevy commented:
@vjeux can we use https://github.com/facebook/fbjs/blob/master/scripts/node/check-dev-engines.js

vjeux commented:
We already have https://github.com/facebook/react-native/blob/master/local-cli/server/checkNodeVersion.js

skevy commented:
@vjeux ah so i guess just exit there if less than node 4.

vjeux commented:
Cool, i'll send a pull request

bestander commented:
Be careful with version checks though.
Remember good old browser sniffing or why there is now Windows 9?

On Thursday, 4 February 2016, Adam Miskiewicz <notifications@github.com>
wrote:

> @vjeux <https://github.com/vjeux> ah so i guess just exit there if less
> than node 4.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react-native/pull/5214#issuecomment-180070095>
> .
>

lrettig commented:
Is there an ETA on this? Thanks!

skevy commented:
@lrettig this is already merged in master. It should be in the current RC and will be officially released in 0.20.

lrettig commented:
Thanks. Didn't see it mentioned in the release notes!

On Thu, Feb 11, 2016 at 12:02 PM, Adam Miskiewicz <notifications@github.com>
wrote:

> @lrettig <https://github.com/lrettig> this is already merged in master.
> It should be in the current RC and will be officially released in 0.20.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react-native/pull/5214#issuecomment-182958680>
> .
>
